src/Lean/Environment.lean

```diff
@@ -1993,6 +1993,9 @@ def hasUnsafe (env : Environment) (e : Expr) : Bool :=
 def realizeConst (env : Environment) (forConst : Name) (constName : Name)
     (realize : Environment → Options → BaseIO (Environment × Dynamic)) :
     IO (Environment × Dynamic) := do
+  -- the following code is inherently non-deterministic in number of heartbeats, reset them at the
+  -- end
```

Comment:
Can this still hit a heartbeat limit in the middle of this nondeterministic phase?

Reply:
It can inside `realize`, but whether that happens is again deterministic fortunately https://github.com/leanprover/lean4/blob/0f3d426591ebcb63f1c38a3074fb2a2ebbae447f/src/Lean/Meta/Basic.lean#L2278